### PR TITLE
fix Issue 22303 - ImportC: pragma directives should be ignored

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by WalterBright** [comment]
Spec pull: https://github.com/dlang/dlang.org/pull/3099

**Review by thewilsonator — DISMISSED** [review]
This doesn't fix the issue listed. It explicitly mentions `#pragma STDC whatever` that must be handled properly, also things like `#pragma pack` affect data layout (!!!!) which very much affects program correctness, especially for interoperability.

**Comment by WalterBright** [comment]
`#pragma STDC FP_CONTRACT` provides a way to disallow contracted expressions. ImportC doesn't do contracted expressions anyway, so this can be safely ignored. C11 6.5-8

`#pragma STDC FENV_ACCESS` informs the compiler when code can manipulate the floating point modes and flags. DMD's code generator does not have support for this, so setting it does not help. This is an issue, but a pretty low priority one, and would be a separate bugzilla from this one.

`#pragma STDC CX_LIMITED_RANGE` are for the usual mathematical formulas for complex arithmetic. I'm pretty sure we use the accurate formulas anyway, so this is irrelevant for us and ignoring it is ok.

`#pragma pack` is the old, non-Standard pre C11 way. The C11 way is to use `_Alignof` and `_Alignas`, which ImportC supports. In general, ImportC is not going to support the endless variety of C extensions, especially ones that now have C11 blessed ways of doing it. C11 is 10 years old now. No excuse for using `pack` anymore.

**Comment by WalterBright** [comment]
https://issues.dlang.org/show_bug.cgi?id=22305

**Comment by thewilsonator** [comment]
In which case you should issue an error for usage of `#pragma pack`

File: src/dmd/lexer.d [hunk 1]
```diff
@@ -1146,6 +1146,11 @@ class Lexer
                             poundLine(n, false);
                             continue;
                         }
+                        else if (n.ident == Id.__pragma)
+                        {
+                            pragmaDirective();
+                            continue;
+                        }
```
Comment (ibuclaw) [inline src/dmd/lexer.d:1153]: Wouldn't this add #pragma to D parser?

**Comment by DaveP1776** [comment]
I filed the bug. The STDC pragmas were to demonstrate that C11 requires processing of pragmas that can’t be removed by the preprocessor. None of the code I was testing importC on actually used those pragmas.

File: src/dmd/lexer.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/lexer.d:1153]: Indeed it would. Oops!

**Comment by WalterBright** [comment]
> In which case you should issue an error for usage of #pragma pack

Some C compilers use `#pragma pack`, others use `#pragma align`. All of them just ignore the ones they don't understand. ImportC will ignore both. This mess is why C11 added `_Alignas`. In order to make ImportC achievable in bounded time, we can't emulate every C extension out there. This is why we're going with C11 6.10.6-1 which sez "Any such pragma that is not recognized by the implementation is ignored."

File: src/dmd/lexer.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/lexer.d]: fixed.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**